src/cargo/sources/registry/http_remote.rs

```diff
@@ -0,0 +1,1209 @@
+
+        // If the path is already being fetched, don't fetch it again.
+        // Just note down the version requirement and move on.
+        if let Some(token) = self.downloads.pending_ids.get(path) {
+            let (dl, _) = self
+                .downloads
+                .pending
+                .get_mut(token)
+                .expect("invalid token");
+
+            trace!("amending dependency that we're already fetching: {}", name);
+            if let Some(req) = req {
+                trace!("adding req {}", req);
+                dl.reqs.insert(req.clone());
+            }
+            dl.is_transitive &= is_transitive;
+
+            return Ok(());
+        } else if self.fresh.contains(path) {
+            // This must have been a 404 when we initially prefetched it.
+            return Ok(());
+        } else if let Some(f) = self.downloads.eager.get_mut(path) {
+            // We can't hit this case.
+            // The index file must exist for the path to be in `eager`,
+            // but since that's the case, we should have caught this
+            // in the eager check _in_ the pkg.exists() path.
+            unreachable!(
+                "index file `{}` is in eager, but file doesn't exist",
+                f.path.display()
+            );
+        }
+
+        if was.is_some() {
+            debug!("double-checking freshness of {}", path.display());
+        }
+
+        // Looks like we're going to have to bite the bullet and do a network request.
+        let url = self.source_id.url();
+        self.prepare()?;
+
+        let mut handle = ops::http_handle(self.config)?;
+        debug!("prefetch {}{}", url, path.display());
+        handle.get(true)?;
+        handle.url(&format!("{}{}", url, path.display()))?;
+        handle.follow_location(true)?;
+
+        // Enable HTTP/2 if possible.
+        if self.multiplexing {
+            try_old_curl!(handle.http_version(HttpVersion::V2), "HTTP2");
+        } else {
+            handle.http_version(HttpVersion::V11)?;
+        }
+
+        // This is an option to `libcurl` which indicates that if there's a
+        // bunch of parallel requests to the same host they all wait until the
+        // pipelining status of the host is known. This means that we won't
+        // initiate dozens of connections to crates.io, but rather only one.
+        // Once the main one is opened we realized that pipelining is possible
+        // and multiplexing is possible with static.crates.io. All in all this
+        // reduces the number of connections done to a more manageable state.
+        try_old_curl!(handle.pipewait(true), "pipewait");
+
+        // Make sure we don't send data back if it's the same as we have in the index.
+        if let Some((ref etag, ref last_modified, _)) = was {
+            let mut list = List::new();
+            list.append(&format!("If-None-Match: {}", etag))?;
+            list.append(&format!("If-Modified-Since: {}", last_modified))?;
```

Comment:
Should this handle servers which don't send both of these headers in the initial response? I'm not sure if we're guaranteed to have both to send back?